src/webui/api/torrentscontroller.cpp

```diff
@@ -1475,6 +1475,33 @@ void TorrentsController::addTagsAction()
     }
 }
 
+void TorrentsController::setTagsAction()
+{
+    requireParams({u"hashes"_s, u"tags"_s});
+
+    const QStringList hashes {params()[u"hashes"_s].split(u'|')};
+    const QStringList tags {params()[u"tags"_s].split(u',', Qt::SkipEmptyParts)};
+
+    // Convert QStringList to TagSet
+    TagSet newTags;
+    for (const QString &tagStr : tags)
+        newTags.insert(Tag(tagStr));
+
+    // Apply the new tags to the selected torrents
+    applyToTorrents(hashes, [&newTags](BitTorrent::Torrent *const torrent)
+    {
+        // Identify tags to add
+        for (const Tag &tag : newTags)
+            if (!torrent->hasTag(tag))
+                torrent->addTag(tag);
+
+        // Identify tags to remove
+        for (const Tag &tag : asConst(torrent->tags()))
+            if (!newTags.contains(tag))
+                torrent->removeTag(tag);
+    });
```

Comment:
why not just remove all tags first and then add new tags?
```suggestion
    applyToTorrents(hashes, [&tags](BitTorrent::Torrent *const torrent)
    {
        torrent->removeAllTags();
        for (const Tag &tag : newTags)
            torrent->addTag(tag);
    });
```

Reply:
Or the following version if the side-effects are bad.
```suggestion
    const TagSet newTags {tags.begin(), tags.end()};
    applyToTorrents(hashes, [&newTags](BitTorrent::Torrent *const torrent)
    {
        for (const Tag &tag : asConst(torrent->tags()))
            if (!newTags.contains(tag))
                torrent->removeTag(tag);
        for (const Tag &tag : newTags)
            torrent->addTag(tag);
    });
```

Reply:
`removeAllTags` just loops over the tags and calls `removeTag` so it would indeed be the same thing as us doing that loop and calling `removeTag`.

However, I think in most cases the torrents will have at least one tag and we add additional tags or remove tags so doing the contains check will keep some of them so it doesn't have to remove X. Each call to `torrentimpl->addTag` or `removeTag` triggers a deferred resume data call and a call to the session to notify of updates.

With that said, I think the latter option to keep the contains check is the preferred way but open for feedback!